sky/provision/azure/instance.py

```diff
@@ -45,64 +141,412 @@ def get_azure_sdk_function(client: Any, function_name: str) -> Callable:
+                f' {verb} still in STOPPING state on Azure. It can only be '
+                'resumed after it is fully STOPPED. Waiting ...')
+        while (stopping_instances and
+               to_start_count > len(stopped_instances) and
+               time.time() - time_start < _RESUME_INSTANCE_TIMEOUT):
+            inst = stopping_instances.pop(0)
+            per_instance_time_start = time.time()
+            while (time.time() - per_instance_time_start <
+                   _RESUME_PER_INSTANCE_TIMEOUT):
+                status = _get_instance_status(compute_client, inst,
+                                              resource_group)
+                if status == AzureInstanceStatus.STOPPED:
+                    break
+                time.sleep(1)
+            else:
+                logger.warning(
+                    f'Instance {inst.name} is still in stopping state '
+                    f'(Timeout: {_RESUME_PER_INSTANCE_TIMEOUT}). '
+                    'Retrying ...')
+                stopping_instances.append(inst)
+                time.sleep(5)
+                continue
+            stopped_instances.append(inst)
+        if stopping_instances and to_start_count > len(stopped_instances):
+            msg = ('Timeout for waiting for existing instances '
+                   f'{stopping_instances} in STOPPING state to '
+                   'be STOPPED before restarting them. Please try again later.')
+            logger.error(msg)
+            raise RuntimeError(msg)
+
+        resumed_instances = stopped_instances[:to_start_count]
+        resumed_instances.sort(key=lambda x: x.name)
+        resumed_instance_ids = [t.name for t in resumed_instances]
+        logger.debug(
+            f'run_instances: Resuming stopped instances {resumed_instance_ids}.'
+        )
+        start_virtual_machine = _get_azure_sdk_function(
+            compute_client.virtual_machines, 'start')
+        with pool.ThreadPool() as p:
+            p.starmap(
+                start_virtual_machine,
+                [(resource_group, inst.name) for inst in resumed_instances])
+
+    to_start_count -= len(resumed_instance_ids)
+
+    if to_start_count > 0:
+        resource_client = azure.get_client('resource', subscription_id)
+        logger.debug(f'run_instances: Creating {to_start_count} instances.')
+        created_instances = _create_instances(
+            compute_client=compute_client,
+            resource_client=resource_client,
+            cluster_name_on_cloud=cluster_name_on_cloud,
+            resource_group=resource_group,
+            provider_config=provider_config,
+            node_config=config.node_config,
+            tags=tags,
+            count=to_start_count)
+        created_instance_ids = [inst.name for inst in created_instances]
+
+    non_running_instances = list(
+        set(AzureInstanceStatus) - {AzureInstanceStatus.RUNNING})
+    while True:
```

Comment:
should we have a timeout here to avoid an infinite loop if Azure fails to start the instances (say quota issues)?